hw/src/lib.rs

```diff
@@ -292,10 +328,75 @@ impl HardwareWalletManager {
 
 impl Drop for HardwareWalletManager {
 	fn drop(&mut self) {
-		// Indicate to the USB Hotplug handlers that they
-		// shall terminate but don't wait for them to terminate.
-		// If they don't terminate for some reason USB Hotplug events will be handled
+		// Indicate to the USB Hotplug handler that it
+		// shall terminate but don't wait for it to terminate.
+		// If it doesn't terminate for some reason USB Hotplug events will be handled
 		// even if the HardwareWalletManger has been dropped
 		self.exiting.store(true, atomic::Ordering::Release);
 	}
 }
+
+/// Hardware wallet event handler
+///
+/// Note, that this run to completion and race-conditions can't occur but this can
+/// therefore starve other events for being process with a spinlock or similar
+struct EventHandler {
+	ledger: Weak<ledger::Manager>,
+	trezor: Weak<trezor::Manager>,
+}
+
+impl EventHandler {
+	/// Trezor event handler constructor
+	pub fn new(ledger: Weak<ledger::Manager>, trezor: Weak<trezor::Manager>) -> Self {
+		Self { ledger, trezor }
+	}
+
+	fn extract_device_info(device: &libusb::Device) -> Result<(u16, u16), Error> {
+		let desc = device.device_descriptor()?;
+		Ok((desc.vendor_id(), desc.product_id()))
+	}
+}
+
+impl libusb::Hotplug for EventHandler {
+	fn device_arrived(&mut self, device: libusb::Device) {
+		// Upgrade reference to an Arc
+		if let (Some(ledger), Some(trezor)) = (self.ledger.upgrade(), self.trezor.upgrade()) {
```

Comment:
q: is it impossible to have one upgrade call return `Some` and the other `None`? Wouldn't that cause the whole block to be skipped? In other words, are both `upgrade()` calls going to return `Some` if one of them does?

Reply:
Good question, because the HardwareWalletManager keeps an `Arc` for both `Ledger` and `Trezor` for the entire lifetime of Parity unless it is disabled by the cli arg `--no-hardware-wallet`! upgrade()` will always succeed as long the `HardwareWalletManager` is enabled (the strong count will be at least 1). So, either both `Trezor` and `Ledger` is enabled or none is enabled.

However, I admit it is a bit sloppy. Theoretically, interleaving "could" happen if one of the `hw-wallets` has been dropped and the other has not been dropped but this is matter in granularity in terms of microseconds and I don't worry too much about this.

I can change it if you like? :P